src/black/linegen.py

```diff
@@ -857,24 +878,34 @@ def normalize_invisible_parens(
             elif not (isinstance(child, Leaf) and is_multiline_string(child)):
                 wrap_in_parentheses(node, child, visible=False)
 
-        check_lpar = isinstance(child, Leaf) and child.value in parens_after
+        comma_check = child.type == token.COMMA if preview else False
+
+        check_lpar = (
+            isinstance(child, Leaf) and child.value in parens_after or comma_check
```

Comment:
```suggestion
            child.value in parens_after if isinstance(child, Leaf) else comma_check
```

Reply:
@JelleZijlstra I don't think this change is right. A comma is a `Leaf` as well so the comma wouldn't get checked here. The comma check doesn't need the `Leaf` check since it uses `.type` so I can do either of these if it's more clear?
```python
isinstance(child, Leaf) and (child.value in parens_after or comma_check)
```
or
```python
comma_check or (isinstance(child, Leaf) and (child.value in parens_after or comma_check)
```

Reply:
Thanks! Yes I assumed that `x and y or z` is equivalent to `y if x else z`, but that's not always true. Parens would make it more clear.